Automattic/sensei
# [POC] Add email data preparation and placeholder replacer functionality

Fixes https://github.com/Automattic/sensei/issues/6461

### Changes proposed in this Pull Request

* Added functionality to replace placeholders in email templates
* Added hooks for extending the ability to send any email with any placeholder or data from anywhere.

### Testing instructions
- Run composer install
- Enable the email post type following the direction here https://github.com/Automattic/sensei/pull/6487
- Change the recipient email using the following hook, or change a teacher's email to something you can check with (You can also use mailhog or something like that)
```
add_filter( 'sensei_test_email_address', function() {
	return 'imranthesuperman@justiceleague.com';
} );
```
- Go to the Gutenberg editor for creating a post of the new post type
- Copy and paste the following markup in the editor
```



<div class="wp-block-group has-text-color has-background" style="color:#404040;background-color:#e6e6e6;padding-top:40px;padding-right:40px;padding-bottom:40px;padding-left:40px">
<p class="has-text-color" style="color:#0b0b0b;font-size:44px;font-style:normal;font-weight:600">[student:displayname]</p>



<p class="has-text-color" style="color:#030303"><strong>Course Name</strong><br><sub>[course.name]</sub></p>



<div class="wp-block-buttons">
<div class="wp-block-button has-custom-font-size has-x-small-font-size" style="font-style:normal;font-weight:400;text-transform:capitalize"><a class="wp-block-button__link has-text-color has-background wp-element-button" style="border-radius:3px;color:#fafafa;background-color:#090909;padding-top:14px;padding-right:20px;padding-bottom:14px;padding-left:20px">Manage students</a></div>
</div>
</div>

```
- Save the post and add meta '_sensei_email_type' = 'student_started_course_to_teacher' . You can use the following snippet `update_post_meta( $email_id, '_sensei_email_type', 'student_started_course_to_teacher' );`
- Create a course with the aforementioned teacher as the author
- Enroll in that course as a student
- A mail should be sent, or you can go to Student Management window and reset the progress to resend the email

### Screenshot / Video
## Editor
![image](https://user-images.githubusercontent.com/6820724/218098604-a81b6c6e-3319-46b7-acb9-843c5d99b965.png)

## Email in client after replacing
![image](https://user-images.githubusercontent.com/6820724/218098766-50a0b947-2d61-4f52-8e78-081a198abd77.png)
